### PR TITLE
Add dev.neovide.neovide

### PR DESCRIPTION
**Pull request by gmankab**
### Please confirm your submission meets all the criteria

- [x] Please describe the application briefly.
This is a simple, no-nonsense, cross-platform graphical user interface for [Neovim](https://github.com/neovim/neovim) (an aggressively refactored and updated Vim editor). Where possible there are some graphical improvements, but functionally it should act like the terminal UI.

i forked @sqwxl's manifest - https://github.com/flathub/flathub/pull/5017

what i did:
- building copy of neovim within the flatpak
- updated sdk to 24.08
- use latest upstream neovide release
- added host-spawn binary, needed to call host shell from neovim sandbox (`:term host-spawn`)
- wrote readme.md with giude on how to tell neovide to use neovim from host, from toolbox, from distrobox, and how to get back to flatpak sandboxed neovim

- [x] The domain used for the application ID is [controlled by the application developers][appid-domain] and the [application id guidelines][appid] are followed.
- [x] I have read the and followed all the [Submission and licence requirements][reqs].
- [x] I have [built][build] and tested the submission locally.
- [x] I am an author/developer/upstream contributor of the project. If not, I contacted upstream developers about this submission. Link: https://github.com/neovide/neovide/pull/2397

[appid-domain]: https://docs.flathub.org/docs/for-app-authors/requirements/#control-over-domain-or-repository
[appid]: https://docs.flathub.org/docs/for-app-authors/requirements#application-id
[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[build]: https://docs.flathub.org/docs/for-app-authors/submission/#before-submission

### CODE REVIEW

**Comment by gmankab** [comment]
bot, build dev.neovide.neovide

File: dev.neovide.neovide.metainfo.xml [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+      <li>Blurred floating windows</li>
+      <li>Emoji support</li>
+    </ul>
+  </description>
+  <categories>
+    <category>Utility</category>
+    <category>TextEditor</category>
+  </categories>
+  <launchable type="desktop-id">dev.neovide.neovide.desktop</launchable>
+  <icon type="stock">dev.neovide.neovide</icon>
```
Comment (bbhtt) [inline dev.neovide.neovide.metainfo.xml:26]:
```suggestion
```

File: dev.neovide.neovide.metainfo.xml
Comment (bbhtt) [inline dev.neovide.neovide.metainfo.xml:1]: this should be submitted upstream https://github.com/neovide/neovide/

File: dev.neovide.neovide.yml [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+runtime: org.freedesktop.Platform
+runtime-version: "24.08"
+sdk: org.freedesktop.Sdk
+command: neovide
+finish-args:
+  - --socket=wayland
+  - --socket=fallback-x11
+  - --share=ipc
+  - --device=dri
+  - --talk-name=org.freedesktop.Flatpak
```
Comment (bbhtt) [inline dev.neovide.neovide.yml:11]: why is this needed? This is not allowed by default.

File: dev.neovide.neovide.yml [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+  - --filesystem=host
+  - --filesystem=/tmp
+  - --filesystem=/var/tmp
```
Comment (bbhtt) [inline dev.neovide.neovide.yml:14]: why?

File: .github/workflows/build-neovide.yml [hunk 4]
```diff
@@ -0,0 +1,17 @@
+name: build flatpak
+on: [push, pull_request]
+jobs:
+  build:
+    name: build flatpak
+    runs-on: ubuntu-latest
```
Comment (bbhtt) [inline .github/workflows/build-neovide.yml:6]: why?

File: dev.neovide.neovide.yml (hunk 2 above)
Comment (bbhtt) [reply dev.neovide.neovide.yml:11]: Oh, hmm the original exception from the previous PR is still there https://github.com/flathub-infra/flatpak-builder-lint/pull/336/

**Comment by bbhtt** [comment]
Was this https://github.com/flathub/flathub/pull/5017#issuecomment-1989781021 solved?

**Comment by gmankab** [comment]
> Was this [#5017 (comment)](https://github.com/flathub/flathub/pull/5017#issuecomment-1989781021) solved?

yes, neovim is shipped within the flatpak now

File: .github/workflows/build-neovide.yml (hunk 4 above)
Comment (gmankab) [reply .github/workflows/build-neovide.yml:6]: > why?

1. make sure everything successfully building outside of my computer

2. it's very convenient to see a green check mark when everything is ok, and a red cross when everything is broken

it's not necessary, i can remove if you want

File: dev.neovide.neovide.yml (hunk 2 above)
Comment (gmankab) [reply dev.neovide.neovide.yml:11]: `--talk-name=org.freedesktop.Flatpak` needed to run host neovim instance (see readme.md), and to acess host shell from neovim terminal emulator (`:term host-spawn`)

org.vim.Vim and com.vscodium.codium packages also uses this permission for terminal emulating


and about fallback-x11, ipc, dri permissions, i don't really know why this permissions needed for, this manifest was originally written by @sqwxl, so probably we should ask them, i didn't touch any permissions

File: dev.neovide.neovide.yml (hunk 3 above)
Comment (gmankab) [reply dev.neovide.neovide.yml:14]: same as above, written by @sqwxl

File: dev.neovide.neovide.metainfo.xml
Comment (gmankab) [reply dev.neovide.neovide.metainfo.xml:1]: there is a pr https://github.com/neovide/neovide/pull/2413

File: .github/workflows/build-neovide.yml (hunk 4 above)
Comment (bbhtt) [reply .github/workflows/build-neovide.yml:6]: the buildbot is going to do that for you, remove this.

File: dev.neovide.neovide.yml (hunk 3 above)
Comment (bbhtt) [reply dev.neovide.neovide.yml:14]: Did you verify if this is needed? If not then remove it.

**Comment by bbhtt** [comment]
you are using a different ID, `neovide` in small caps. Can you make a PR to change that here https://github.com/flathub-infra/flatpak-builder-lint/blob/03ecc66c93c86eb8913b6e658206b06472c16864/flatpak_builder_lint/staticfiles/exceptions.json#L2403

File: flathub.json [hunk 5]
```diff
@@ -0,0 +1,5 @@
+{
+  "only-arches": [
+    "x86_64"
```
Comment (hfiguiere) [inline flathub.json:3]: why do you disallow this from running on a RaspberryPi?

File: dev.neovide.neovide.yml [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+        sha256: 75159f0d3924a0bd6d230c6414d19ad44609b5a16671db6624681c910f9bb4ff
+        dest-filename: host-spawn
+  - name: neovide
+    sources:
+      - type: git
+        url: https://github.com/neovide/neovide
+        commit: 9c23c432b8105f39d9e9ddc2732d3350a0323cc9
+        dest: repo
+      - type: archive
+        url: https://github.com/neovide/neovide/releases/download/0.13.3/neovide-linux-x86_64.tar.gz
```
Comment (hfiguiere) [inline dev.neovide.neovide.yml:221]: why not building from source?

File: dev.neovide.neovide.yml [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+          type: anitya
+          project-id: 9037
+          url-template: https://github.com/neovim/neovim/archive/v$version.tar.gz
+  - name: host-spawn
+    buildsystem: simple
+    build-commands:
+      - install -Dm755 host-spawn /app/bin/host-spawn
+    sources:
+      - type: file
+        url: https://github.com/1player/host-spawn/releases/download/v1.6.0/host-spawn-x86_64
```
Comment (hfiguiere) [inline dev.neovide.neovide.yml:211]: why not building this from source?

File: flathub.json (hunk 5 above)
Comment (gmankab) [reply flathub.json:3]: yea, this should be removed

File: dev.neovide.neovide.yml (hunk 6 above)
Comment (gmankab) [reply dev.neovide.neovide.yml:221]: i spend about 10 hours trying to build neovide in flatpak sandbox without network access, without success

feel free to submit changes

File: dev.neovide.neovide.yml (hunk 7 above)
Comment (gmankab) [reply dev.neovide.neovide.yml:211]: copied this part from vscodium manifest

i'll try building from source

**Comment by gmankab** [comment]
> you are using a different ID, `neovide` in small caps. Can you make a PR to change that here https://github.com/flathub-infra/flatpak-builder-lint/blob/03ecc66c93c86eb8913b6e658206b06472c16864/flatpak_builder_lint/staticfiles/exceptions.json#L2403?

done - https://github.com/flathub-infra/flatpak-builder-lint/pull/515

File: dev.neovide.neovide.yml (hunk 3 above)
Comment (gmankab) [reply dev.neovide.neovide.yml]: removed https://github.com/flathub/flathub/pull/5745/commits/d5b40c8a74f0a438f5a3faccaf1a80e907f19534

File: .github/workflows/build-neovide.yml (hunk 4 above)
Comment (gmankab) [reply .github/workflows/build-neovide.yml]: removed https://github.com/flathub/flathub/pull/5745/commits/e3e915db51b27f1f48c791b5edca2830490b8cbe

**Comment by gmankab** [comment]
bot, build dev.neovide.neovide

**Comment by gmankab** [comment]
bot, build dev.neovide.neovide

**Comment by gmankab** [comment]
bot, build dev.neovide.neovide

File: dev.neovide.neovide.yml (hunk 6 above)
Comment (gmankab) [reply dev.neovide.neovide.yml]: fixed, now building from source

File: dev.neovide.neovide.yml (hunk 7 above)
Comment (gmankab) [reply dev.neovide.neovide.yml]: fixed, now building from source

**Comment by gmankab** [comment]
related - https://github.com/rust-skia/rust-skia/issues/962

File: modules/host-spawn/host-spawn.yml
Comment (hfiguiere) [inline modules/host-spawn/host-spawn.yml:1]: the content of this file should inline in the manifest. it's not autogenerated.

File: modules/neovide/neovide.yml
Comment (hfiguiere) [inline modules/neovide/neovide.yml:1]: the content of this file should be inlione in the manifest, it's not autogenerated.

File: modules/neovim.yml
Comment (hfiguiere) [inline modules/neovim.yml:1]: the content of this file should be inlione in the manifest, it's not autogenerated.

File: dev.neovide.neovide.yml [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+  - --socket=fallback-x11
+  - --share=ipc
+  - --device=dri
+  - --talk-name=org.freedesktop.Flatpak
+  - --filesystem=host
+  - --share=network
+modules:
+  - modules/neovim.yml
+  - modules/host-spawn/host-spawn.yml
+  - modules/neovide/neovide.yml
```
Comment (hfiguiere) [inline dev.neovide.neovide.yml:17]: and just to reiterate, these non autogenerated portion of the manifest should be inline here like they were before.

File: modules/neovim.yml
Comment (hfiguiere) [reply modules/neovim.yml:1]: no it's not easier. It's way harder that way.

File: modules/neovide/dev.neovide.neovide.metainfo.xml
Comment (bbhtt) [inline modules/neovide/dev.neovide.neovide.metainfo.xml:1]: was this submitted upstream?

File: modules/neovim.yml [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+      - PREFIX=/app
+    sources:
+      - type: archive
+        url: http://www.leonerd.org.uk/code/libtermkey/libtermkey-0.22.tar.gz
+        sha256: 6945bd3c4aaa83da83d80a045c5563da4edd7d0374c62c0d35aec09eb3014600
+    cleanup:
+      - /include
+      - /lib/pkgconfig
+      - /share/man
+      - '*.la'
```
Comment (bbhtt) [inline modules/neovim.yml:162]: Add these as a global cleanup in the main manifest, no need to repeat it for everything and make manifests larger.

Comment (bbhtt) [reply modules/neovim.yml:162]: "*.a", "*.la" etc. too

File: modules/neovide/neovide.yml [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+  append-path: /usr/lib/sdk/rust-stable/bin
+  env:
+    CARGO_HOME: /run/build/neovide/cargo
+    SKIA_BINARIES_URL: "file:///run/build/neovide/skia-binaries.tar.gz"
+    FORCE_SKIA_BINARIES_DOWNLOAD: "1"
+buildsystem: simple
+build-commands:
+  - cargo --frozen fetch
+  - cargo --frozen build --release
+  - ls target
```
Comment (bbhtt) [inline modules/neovide/neovide.yml:31]:
```suggestion
```

File: modules/neovide/dev.neovide.neovide.metainfo.xml
Comment (gmankab) [reply modules/neovide/dev.neovide.neovide.metainfo.xml:1]: https://github.com/neovide/neovide/pull/2413

File: modules/neovim.yml (hunk 9 above)
Comment (gmankab) [reply modules/neovim.yml:162]: @bbhtt, cleanup started with * is meant to be used on module level

are you sure this should be moved to the main manifest too?

Comment (hfiguiere) [reply modules/neovim.yml:162]: `*.a` and `*.la` never have any use in a flatpak. @bbhtt is correct.

File: modules/neovim.yml
Comment (hfiguiere) [reply modules/neovim.yml:1]: no it's not. You can't even see at first glance what's happening. I guess you don't maintain 50+ packages...

File: modules/neovim.yml (hunk 9 above)
Comment (gmankab) [reply modules/neovim.yml]: commited - https://github.com/flathub/flathub/pull/5745/commits/d578c94b5c4767c4ebaa1f09737e884bf4423957

can you please check did i do it correctly?

File: modules/neovim.yml
Comment (gmankab) [reply modules/neovim.yml:1]: maybe it's easier for you to maintain one big file, but i am sure that it's not easier for me

i understand that you have much more experience in maintaining manifests, but your personal experience does not apply to me

unless you have a strict requirement that everything be in one file, i prefer modular structure with multiple files, which is more comfortable for me

**Comment by gmankab** [comment]
> related - [rust-skia/rust-skia#962](https://github.com/rust-skia/rust-skia/issues/962)

this issue says that clang17 needed to solve problem that we getting while building neovide for arm

but org.freedesktop.Sdk.Extension.llvm17 is only available for 23.08 runtime, and not available for 24.08

so we should choose one of this vaiants:

- temporary drop arm support untill issue solved in upstream
- port llvm17 to 24.08 runtime and use it for building

i am not really interested in working with legacy llvm17, so temporary drop arm support seems more reasonable for me

@hfiguiere, what do you think?

**Comment by bbhtt** [comment]
> port llvm17 to 24.08 runtime and use it for building

LLVM 17 is end of life so that's not going to happen. Does it work with llvm 18+?

**Comment by gmankab** [comment]
i tried llvm 18 and 19, both work on x86, and not wok on arm

also i tried downgrade to 23.04 sdk and building with llvm17, and i am getting same error, despite the fact that issue thread linked above says llvm17 should work

maybe i am doing something wrong

**Comment by gmankab** [comment]
bot, build dev.neovide.neovide

**Comment by gmankab** [comment]
bot, build dev.neovide.neovide

**Comment by bbhtt** [comment]
bot, build dev.neovide.neovide

**Comment by bbhtt** [comment]
let us know when this is ready to review again.

**Comment by gmankab** [comment]
> let us know when this is ready to review again.

this is ready (probably)

i bought xiaomi pad 5, i am planning to install postmarket os and investigate problems with building on arm

there is a neovide build for arch linux arm, so they solved building problem somehow

**Comment by bbhtt** [comment]
bot, build dev.neovide.neovide

File: dev.neovide.neovide.yml [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+  - --talk-name=org.freedesktop.Flatpak
+  - --filesystem=host
+  - --share=network
+  - --env=TERM=xterm-256color
+modules:
+  - modules/host-spawn/host-spawn.yml
+  - modules/fd/fd.yml
+  - modules/ripgrep/ripgrep.yml
+  - modules/neovim/neovim.yml
+  - modules/neovide/neovide.yml
```
Comment (bbhtt) [inline dev.neovide.neovide.yml:23]: i'd prefer if you have neovide in the main manifest as that's the main module.

Comment (gmankab) [reply dev.neovide.neovide.yml]: > i'd prefer if you have neovide in the main manifest as that's the main module.

i don't like it, but ok, done

**Comment by bbhtt** [comment]
bot, build dev.neovide.neovide

**Comment by gmankab** [comment]
thank you
